<a href="https://colab.research.google.com/github/Hina19/Chest_ray/blob/master/07-08-09-Pandas/PostRead-Pandas2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

####Dealign with Missing values in IMDB dataset

Before that lets do some exploration on our movies dataset before jumping into dealing with missing values



!wget "https://drive.google.com/uc?export=download&id=1UoooIoWfUFswDuPndOpHOLYE52dSRwCh" -O data.csv



In [ ]:
!wget "https://drive.google.com/uc?export=download&id=1UoooIoWfUFswDuPndOpHOLYE52dSRwCh" -O data.csv

--2022-05-16 13:14:27--  https://drive.google.com/uc?export=download&id=1UoooIoWfUFswDuPndOpHOLYE52dSRwCh
Resolving drive.google.com (drive.google.com)... 173.194.216.138, 173.194.216.102, 173.194.216.101, ...
Connecting to drive.google.com (drive.google.com)|173.194.216.138|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-04-68-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/h6qvgq9ghj3bhktbdniukkr2v05ni1u8/1652706825000/14302370361230157278/*/1UoooIoWfUFswDuPndOpHOLYE52dSRwCh?e=download [following]
--2022-05-16 13:14:27--  https://doc-04-68-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/h6qvgq9ghj3bhktbdniukkr2v05ni1u8/1652706825000/14302370361230157278/*/1UoooIoWfUFswDuPndOpHOLYE52dSRwCh?e=download
Resolving doc-04-68-docs.googleusercontent.com (doc-04-68-docs.googleusercontent.com)... 74.125.31.132, 2607:f8b0:400c:c02::84
Connecting to doc-04-68-docs.googleusercontent.com (doc-04-68-d

In [ ]:
data = pd.read_csv('data.csv',index_col=0)

In [ ]:
data


id  popularity                                     title  \
0     43597         150                                    Avatar   
1     43598         139  Pirates of the Caribbean: At World's End   
2     43599         107                                   Spectre   
3     43600         112                     The Dark Knight Rises   
4     43602         115                              Spider-Man 3   
...     ...         ...                                       ...   
1460  48363           3                            The Last Waltz   
1461  48370          19                                    Clerks   
1462  48375           7                                   Rampage   
1463  48376           3                                   Slacker   
1464  48395          14                               El Mariachi   

      vote_average  vote_count  year month       day      director_name  \
0              7.2       11800  2009   Dec  Thursday      James Cameron   
1              6.9        4500  2007   May  Saturday     Gore Verbinski   
2              6.3        4466  2015   Oct    Monday         Sam Mendes   
3              7.6        9106  2012   Jul    Monday  Christopher Nolan   
4              5.9        3576  2007   May   Tuesday          Sam Raimi   
...            ...         ...   ...   ...       ...                ...   
1460           7.9          64  1978   May    Monday    Martin Scorsese   
1461           7.4         755  1994   Sep   Tuesday        Kevin Smith   
1462           6.0         131  2009   Aug    Friday           Uwe Boll   
1463           6.4          77  1990   Jul    Friday  Richard Linklater   
1464           6.6         238  1992   Sep    Friday   Robert Rodriguez   

     gender  revenue_Mdollars  budget_Mdollars  
0      Male           2787.97           237.00  
1      Male            961.00           300.00  
2      Male            880.67           245.00  
3      Male           1084.94           250.00  
4      Male            890.87           258.00  
...     ...               ...              ...  
1460   Male              0.32             0.00  
1461   Male              3.15             0.03  
1462   Male              0.00             0.00  
1463   Male              0.00             0.00  
1464    NaN              2.04             0.22  

[1465 rows x 12 columns]



- We can check **`data.info()`** to get an **idea of distribution of missing values**


- It gives **no. of non-null (Available) values** in each column

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1465 entries, 0 to 1464
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id_x           1465 non-null   int64  
 1   budget         1465 non-null   float64
 2   popularity     1465 non-null   int64  
 3   revenue        1465 non-null   float64
 4   title          1465 non-null   object 
 5   vote_average   1465 non-null   float64
 6   vote_count     1465 non-null   int64  
 7   year           1465 non-null   int64  
 8   month          1465 non-null   object 
 9   day            1465 non-null   object 
 10  director_name  1465 non-null   object 
 11  gender         1341 non-null   object 
dtypes: float64(3), int64(4), object(5)
memory usage: 148.8+ KB


#### Notice the dataset

- Some of the values are missing - `NaN` in gender

####But other than that is there any other missing values present ?

Lets check for this -

In [ ]:
data.loc[(data['revenue'] == 0.0) | (data['budget'] == 0.0)]

id_x  budget  popularity  revenue              title  vote_average  \
195   43918     0.0          16   104.91       The Campaign           5.6   
271   44050     0.0          25     0.00   The Pink Panther           5.6   
281   44072     0.0          20    43.31           The Edge           6.7   
314   44133    75.0          13     0.00  Anna and the King           6.4   
341   44181     0.0          13     0.00               Wolf           6.0   
...     ...     ...         ...      ...                ...           ...   
1457  48323     0.0           5     0.00         The Mighty           7.1   
1459  48359     0.0           2     0.00  George Washington           6.4   
1460  48363     0.0           3     0.32     The Last Waltz           7.9   
1462  48375     0.0           7     0.00            Rampage           6.0   
1463  48376     0.0           3     0.00            Slacker           6.4   

      vote_count  year month        day       director_name gender  
195          578  2012   Aug   Thursday           Jay Roach   Male  
271          550  2006   Jan  Wednesday          Shawn Levy   Male  
281          349  1997   Sep   Saturday        Lee Tamahori   Male  
314          197  1999   Dec   Thursday        Andy Tennant   Male  
341          216  1994   Jun     Friday        Mike Nichols   Male  
...          ...   ...   ...        ...                 ...    ...  
1457          51  1998   Oct     Friday       Peter Chelsom   Male  
1459          36  2000   Oct     Sunday  David Gordon Green   Male  
1460          64  1978   May     Monday     Martin Scorsese   Male  
1462         131  2009   Aug     Friday            Uwe Boll   Male  
1463          77  1990   Jul     Friday   Richard Linklater   Male  

[227 rows x 12 columns]

#### What you can refer from above ?

Usually,  Missing value in data can appear in 3 ways

    - NaN
    - NAN
    - nan

####But that does not always hold true.

As you can see movies with `zero budget` or `zero revenue` is not possible, that are nothing but the missing values here 

So you have to be intuitive to decide for missing values in the dataset

Lets convert all missing values in columns to nan for better understanding

In [ ]:
data['revenue']=data['revenue'].replace(0, np.nan)
data['budget']=data['budget'].replace(0, np.nan)



#### Why do we need to handle these missing values?

- Because most ML and DS algorithms break when they encounter missing data


- Missing data depreciate the performance of our models


#### Question: Now, How can we deal with missing values? What ideas do you have?

- It is on **case-to-case basis**


- We have to **pick a method based on the dataset and SITUATION**


- We have to **check what will work and what not**


#### Ask yourself: What makes sense and what not?

- **DON'T just remove OR replace with default value OR replace with mean OR anything else blindly**


- The way we choose to deal with missing values can be **easily misleading**


#### So, we need to be very careful about how we choose to deal with missing data

- Use something **simple**


- But it **should make sense in the given situation**


### There are many ways to deal with missing values


#### First, Let's import NaN from Numpy

In [ ]:
from numpy import NaN, NAN, nan

In [ ]:
# One interesting thing about Nan is this:

nan == nan

False

#### Why did it come out be `False`?

#### Can you compare two infinite or non-existent values?

- NO

- So, **be careful** while searching for missing values **using `==`**


#### So, How can we find if a value is missing or not?

- We can use **`is`**


- We can use **Pandas in-built function `isnull()`**

In [ ]:
nan is nan

True

In [ ]:
pd.isnull(nan)

True

#### Now coming back to our data
### How to find the total number of NULL values in each column:


In [ ]:
data.isnull().sum()

id_x               0
budget           137
popularity         0
revenue          196
title              0
vote_average       0
vote_count         0
year               0
month              0
day                0
director_name      0
gender           124
dtype: int64

####Now if you want to check for the percenatge of missing value for each column or features ?

In [ ]:

# len(movies) represents total number of rows
mis_val_per = data.isnull().sum()/len(movies)*100
mis_val_per


id_x              0.000000
budget            9.351536
popularity        0.000000
revenue          13.378840
title             0.000000
vote_average      0.000000
vote_count        0.000000
year              0.000000
month             0.000000
day               0.000000
director_name     0.000000
gender            8.464164
dtype: float64

#### What if you want to check for specific column gender ?

We can simply pass the column name as

In [ ]:
data['gender'].isnull().sum()

124

How to get the frequency of null values using unique()

- We can get **frequency of each value** in a column


- **No. of occurrences of each value** in a column


- Using `value_counts()`

In [ ]:
data['gender'].value_counts()

Male      1309
Female      32
Name: gender, dtype: int64

#### It tells that

- Value Male occurs 1309 times

- Value Female occurs 32 times




#### But its not telling the count of missing values

- Because by **default**, the **parameter `dropna` is set to `True`**


- `dropna=True` means it is **NOT going to count missing values**


- So, **we have to set `dropna=False`** to get the **count of missing values in a column**

In [ ]:
data['gender'].value_counts(dropna=False)

Male      1309
NaN        124
Female      32
Name: gender, dtype: int64

**Handling missing data**:

After finding the missing values, it needs to treat them.

There are many ways to handle missing values in the dataset

### Removing/Dropping the missing values

#### What if you have just 1 or very few rows which has missing data, compared to whole data?

- **Example**: only 10 rows out of 1 million rows having missing values


- We can simply remove those rows or columns using **`drop()`** - already seen





### Replacing missing values with mean

But before making changes to the orginal data, lets again create a copy of it using copy()

In [ ]:
df=data.copy()

####Any()

####Question: How to filter rows that contains atleast one missing value

We can do it using any() method of pandas
 
 - The any() function is used to check whether any element is True, potentially over an axis.

In [ ]:
# Let's check the rows with null values before replace
df[df.isnull().any(axis=1)]

id_x  budget  popularity  revenue                title  vote_average  \
17    43620  180.00          42   372.23   The Golden Compass           5.8   
38    43653  185.00          65   343.47     Star Trek Beyond           6.6   
50    43669  175.00          90   745.00        Suicide Squad           5.9   
53    43672  175.00          44   264.22           Waterworld           5.9   
81    43729  150.00          44   301.00  Wrath of the Titans           5.5   
...     ...     ...         ...      ...                  ...           ...   
1459  48359     NaN           2      NaN    George Washington           6.4   
1460  48363     NaN           3     0.32       The Last Waltz           7.9   
1462  48375     NaN           7      NaN              Rampage           6.0   
1463  48376     NaN           3      NaN              Slacker           6.4   
1464  48395    0.22          14     2.04          El Mariachi           6.6   

      vote_count  director_id  year month       day       director_name  id_y  \
17          1303         4780  2007   Dec   Tuesday         Chris Weitz  4780   
38          2568         4802  2016   Jul  Thursday          Justin Lin  4802   
50          7458         4811  2016   Aug   Tuesday          David Ayer  4811   
53           992         4814  1995   Jul    Friday      Kevin Reynolds  4814   
81          1431         4849  2012   Mar   Tuesday  Jonathan Liebesman  4849   
...          ...          ...   ...   ...       ...                 ...   ...   
1459          36         5231  2000   Oct    Sunday  David Gordon Green  5231   
1460          64         4809  1978   May    Monday     Martin Scorsese  4809   
1462         131         5148  2009   Aug    Friday            Uwe Boll  5148   
1463          77         5535  1990   Jul    Friday   Richard Linklater  5535   
1464         238         5097  1992   Sep    Friday    Robert Rodriguez  5097   

     gender  
17      NaN  
38      NaN  
50      NaN  
53      NaN  
81      NaN  
...     ...  
1459   Male  
1460   Male  
1462   Male  
1463   Male  
1464    NaN  

[325 rows x 14 columns]

- The any() method returns one value for each column, True if ANY value in that column is True, otherwise False.

- Here .any() checks if any value in a column is null


####Now lets say you want to check for specific column say 'revenue'

In [ ]:
# Now check the null values only in revenue column
df.loc[df[["revenue"]].isna().any(axis=1)]


id_x  budget  popularity  revenue                title  vote_average  \
271   44050     NaN          25      NaN     The Pink Panther           5.6   
314   44133    75.0          13      NaN    Anna and the King           6.4   
341   44181     NaN          13      NaN                 Wolf           6.0   
351   44198     NaN           9      NaN           Rollerball           3.4   
374   44245    65.0          19      NaN      Mona Lisa Smile           6.5   
...     ...     ...         ...      ...                  ...           ...   
1453  48294     NaN           6      NaN  She's Gotta Have It           6.1   
1457  48323     NaN           5      NaN           The Mighty           7.1   
1459  48359     NaN           2      NaN    George Washington           6.4   
1462  48375     NaN           7      NaN              Rampage           6.0   
1463  48376     NaN           3      NaN              Slacker           6.4   

      vote_count  director_id  year month        day       director_name  \
271          550         4842  2006   Jan  Wednesday          Shawn Levy   
314          197         5039  1999   Dec   Thursday        Andy Tennant   
341          216         5031  1994   Jun     Friday        Mike Nichols   
351          106         4890  2002   Feb     Friday      John McTiernan   
374          393         4797  2003   Dec     Friday         Mike Newell   
...          ...          ...   ...   ...        ...                 ...   
1453          25         5195  1986   Aug     Friday           Spike Lee   
1457          51         4921  1998   Oct     Friday       Peter Chelsom   
1459          36         5231  2000   Oct     Sunday  David Gordon Green   
1462         131         5148  2009   Aug     Friday            Uwe Boll   
1463          77         5535  1990   Jul     Friday   Richard Linklater   

      id_y gender  
271   4842   Male  
314   5039   Male  
341   5031   Male  
351   4890    NaN  
374   4797   Male  
...    ...    ...  
1453  5195   Male  
1457  4921   Male  
1459  5231   Male  
1462  5148   Male  
1463  5535   Male  

[196 rows x 14 columns]

####Filling null values in revenue using mean:

lets first calculate the mean of column revenue

In [ ]:
m = df['revenue'].mean()
m

165.37985815602832

#### We can use `replace()` to replace the missing values with whatever we want, Let's say mean

Parameters:

- **`to_replace`** - What do I want to replace? 


- **`value`** - With what I want to replace?


- **`inplace=True`** - For permanent change

In [ ]:
df['revenue'].replace(to_replace=NaN, value=m, inplace=True)
df

id_x  budget  popularity      revenue  \
0     43597  237.00         150  2787.970000   
1     43598  300.00         139   961.000000   
2     43599  245.00         107   880.670000   
3     43600  250.00         112  1084.940000   
4     43602  258.00         115   890.870000   
...     ...     ...         ...          ...   
1460  48363     NaN           3     0.320000   
1461  48370    0.03          19     3.150000   
1462  48375     NaN           7   165.379858   
1463  48376     NaN           3   165.379858   
1464  48395    0.22          14     2.040000   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  

[1465 rows x 12 columns]

* Now we can see in above results, the nan values at index 1462 and 1463 is **replaced with mean value** 165.37




- **`value=` can be anything**
    - mean
    - median
    - mode
    - 0
    - min, max
    - ...

Similarly, you can try to fill the missing values for column budget 



### Replace missing values with mode

After checking for numerical values, now lets see how you are going to handle missing values in categorical column `gender'

In [ ]:
df['gender'].value_counts(dropna=False)

Male      1309
NaN        124
Female      32
Name: gender, dtype: int64

In case of categorical we can replace the missing value by the most frequent value for that particular column, 

Here we replace the missing value by Male since the count of `Male` is more than `Female`

In [ ]:
#Take most occured category in that vairable (.mode())
Mode_Category = df['gender'].mode()
type(Mode_Category)

pandas.core.series.Series

Please note : mode() always returns Series even if only one value is returned.

So inorder to get first element of mode() we have to use index[0]

In [ ]:
Mode_Category = df['gender'].mode()[0]
Mode_Category

'Male'

Now replace NAN values with most occured category in actual variable

NOTE: Features having a max number of null values may create bias if replace with the most occurred category.
   

In [ ]:
 df['gender'].replace(to_replace=NaN, value=Mode_Category,inplace=True)
 df['gender'].value_counts(dropna=False)

Male      1433
Female      32
Name: gender, dtype: int64

In [ ]:
df

id_x  budget  popularity      revenue  \
0     43597  237.00         150  2787.970000   
1     43598  300.00         139   961.000000   
2     43599  245.00         107   880.670000   
3     43600  250.00         112  1084.940000   
4     43602  258.00         115   890.870000   
...     ...     ...         ...          ...   
1460  48363     NaN           3     0.320000   
1461  48370    0.03          19     3.150000   
1462  48375     NaN           7   165.379858   
1463  48376     NaN           3   165.379858   
1464  48395    0.22          14     2.040000   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez   Male  

[1465 rows x 12 columns]

## **Quiz :**


1. What will the following code snippet return?

> dataframe.isnull()

a. A count of the missing values in each column

b. A count of the missing values in each row

c. A single boolean value True if the DataFrame contains one or more missing values, or False if there are none.

d. A corresponding dataframe of boolean values, with True in cells that contain missing values, and False in cells that contain valid data


ans:d


2. You're cleaning up a DataFrame with ~1000 observations. You notice that one categorical column contains 512 missing values. What strategy should you employ to deal with these missing values?

a. Drop all rows with missing values

b. Drop the column entirely

c. Replace all missing values with the column mean

d. Replace all missing values with randomly sampled values from this column

ans: b as we have more than 50% of missing data




